packages/langium/src/lsp/language-server.ts

```diff
@@ -89,33 +87,23 @@ export function startLanguageServer(services: LangiumSharedServices): void {
 
 export function addDocumentsHandler(connection: Connection, services: LangiumSharedServices): void {
     const documentBuilder = services.workspace.DocumentBuilder;
+    const mutex = services.workspace.ReadWriteMutex;
     let changeTokenSource: AbstractCancellationTokenSource;
-    let changePromise: Promise<void> | undefined;
 
-    async function onDidChange(changed: URI[], deleted?: URI[]): Promise<void> {
+    function onDidChange(changed: URI[], deleted: URI[]): void {
         changeTokenSource?.cancel();
-        if (changePromise) {
-            await changePromise;
-        }
         changeTokenSource = startCancelableOperation();
-        changePromise = documentBuilder
-            .update(changed, deleted ?? [], changeTokenSource.token)
-            .catch(err => {
-                if (!isOperationCancelled(err)) {
-                    console.error('Error: ', err);
-                }
-            });
+        const token = changeTokenSource.token;
+        mutex.write(() => documentBuilder.update(changed, deleted, token));
```

Comment:
First thought before looking deeper into this: You now lock the whole workspace until the update is completely done, which includes validation. That last phase of the document builder can potentially take very long and thus block all LSP services.

It would be good to verify this: Implement an async validation check that uses `setTimeout` to simulate a long-running process.

Potential solution: extract the validation phase such that it's performed in a _read_ lock after the _write_ is finished.